appshell/appshell_extensions_win.cpp

```diff
@@ -903,34 +959,106 @@ int32 ReadFile(ExtensionString filename, ExtensionString encoding, std::string&
 
                 validationState.data = buffer;
                 validationState.dataLen = dwFileSize;
-                validationState.preserveBOM = false;
+                validationState.preserveBOM = true;
 
                 if (ReadFile(hFile, buffer, dwFileSize, &dwBytesRead, NULL)) {
-                    if (!GetBufferAsUTF8(validationState)) {
-                        error = ERR_UNSUPPORTED_ENCODING;
-                    } else {
-                        contents = std::string(buffer, validationState.dataLen);
-                    }        
-                } else {
+                    contents = std::string(buffer, validationState.dataLen);
+                    if (encoding != L"UTF-8"|| !GetBufferAsUTF8(validationState)) {
+                        std::string detectedCharSet;
+                        try {
+                            if (encoding == L"UTF-8") {
+                                GetCharsetMatch(contents.c_str(), contents.size(), detectedCharSet);
+                            }
+                            else {
+                                std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> conv;
+                                detectedCharSet = conv.to_bytes(encoding);
+                            }
+                            std::transform(detectedCharSet.begin(), detectedCharSet.end(), detectedCharSet.begin(), ::toupper);
```

Comment:
std::toupper is defined for char only, not for string, so we have to use transform.
https://stackoverflow.com/questions/735204/convert-a-string-in-c-to-upper-case